core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/ExcelInputReader.java

```diff
@@ -0,0 +1,457 @@
+        String columnName = columnValueArray[0];
+        String columnValue = columnValueArray[1];
+
+        if (columnMapping.containsKey(columnName)) {
+          excelColumnValueMap.put(columnMapping.get(columnName), columnValue);
+        } else {
+          excelColumnValueMap.put(columnName, columnValue);
+        }
+      }
+    }
+
+    try {
+      for (Schema.Field field : outputSchema.getFields()) {
+        String fieldName = field.getName();
+        if (excelColumnValueMap.containsKey(fieldName)) {
+          builder.convertAndSet(fieldName, excelColumnValueMap.get(fieldName));
+        } else {
+          builder.set(fieldName, NULL);
+        }
+      }
+
+      builder.set(FILE, new Path(fileName).getName());
+      builder.set(SHEET, sheetName);
+
+      if (ifEndRow.equalsIgnoreCase(END)) {
+        trackProcessedFiles(fileName);
+      }
+
+      emitter.emit(builder.build());
+    } catch (Exception e) {
+      switch (excelInputreaderConfig.ifErrorRecord) {
+        case EXIT_ON_ERROR:
+          throw new IllegalStateException("Terminating processing on error : " + e.getMessage());
+        case WRITE_ERROR_DATASET:
+          StructuredRecord.Builder errorRecordBuilder = StructuredRecord.builder(errorRecordSchema);
+          errorRecordBuilder.set(KEY, fileName + "_" + sheetName + "_" + excelRecords[0]);
+          errorRecordBuilder.set(FILE, fileName);
+          errorRecordBuilder.set(SHEET, sheetName);
+          errorRecordBuilder.set(RECORD, inputValue);
+          writeErrorRecordToDataset(errorRecordBuilder.build());
+          break;
+        default:
+          //ignore on error
+          break;
+      }
+
+    }
+
+  }
+
+  /**
+   * Writes proceesed file name to memory table.
+   * @param fileName
+   */
+  private void trackProcessedFiles(String fileName) {
+    KeyValueTable processedFileMemoryTable = batchRuntimeContext.getDataset(excelInputreaderConfig.memoryTableName);
+    processedFileMemoryTable.write(fileName, String.valueOf(System.currentTimeMillis()));
+  }
+
+  /**
+   * Writes error record to error dataset.
+   * @param errorRecord in the form of {@link StructuredRecord}
+   * throws {@link IOException}
+   */
+  private void writeErrorRecordToDataset(StructuredRecord errorRecord) throws IOException {
+    Table errorTable = batchRuntimeContext.getDataset(excelInputreaderConfig.errorDatasetName);
+    errorTable.write(errorRecord);
+  }
+
+  /**
+   * Returns list of all the processed file names which are kept in memory table.
+   * @param batchSourceContext
+   * @return processedFiles
+   */
+  private List<String> getAllProcessedFiles(BatchSourceContext batchSourceContext) {
+    KeyValueTable table = batchSourceContext.getDataset(excelInputreaderConfig.memoryTableName);
+    List<String> processedFiles = new ArrayList<>();
+    try (CloseableIterator<KeyValue<byte[], byte[]>> filesIterable = table.scan(null, null)) {
+      while (filesIterable.hasNext()) {
```

Comment:
Will the key value table keep growing or is there a TTL set on this. After a while this table will become large and it would take a long time to return. We need to be able to set TTL and we should ask for this configuration from user right ?